File: packages/agent/agent/tools/read_concept_history.ts
import { getDb } from "@clara-financas/db";
import { conceptRevisions, concepts, type Bundle } from "@clara-financas/db/schema/knowledge";
import { forTenant } from "@clara-financas/db/tenant-scope";
import { and, desc, eq } from "drizzle-orm";
import { defineTool } from "eve/tools";
import { z } from "zod";

import { notFound } from "../lib/errors";
import { requireTenantCaller } from "../lib/tenant";

/**
 * O histórico de um conceito — a metade que faltava do revert.
 *
 * `save_concept` sempre prometeu: "revert é escrever de volta um corpo
 * antigo". Só que `concept_revisions` era append-only SEM leitor — não havia
 * como ver o corpo antigo para escrevê-lo de volta. A promessa apontava para
 * uma porta que não existia, o mesmo defeito que `create_adjustment` teve um
 * dia.
 *
 * Cada revisão guarda o frontmatter e o corpo completos do estado que ela
 * gravou. Reverter é: ler aqui, escolher a revisão, e chamar `save_concept`
 * com aquele corpo — o revert vira uma revisão nova, e a trilha continua
 * contando a história inteira.
 */
const LIMIT = 20;

export default defineTool({
  description:
    "Reads the revision history of one concept (learned rule, alias, category), newest first, each with the full body it recorded. To revert a concept, pick the revision and call save_concept with that body — the revert becomes a new revision, nothing is erased.",
  inputSchema: z.object({
    conceptId: z
      .string()
      .min(1)
      .describe("The concept path, e.g. 'rules/mercado-sao-jose' — as shown by read_concept."),
    bundle: z.enum(["constitution", "learnings"]).optional().describe("Defaults to learnings."),
    limit: z.number().int().min(1).max(LIMIT).optional(),
  }),
  async execute(input, ctx) {
    const { tenantId } = requireTenantCaller(ctx);
    const bundle: Bundle = input.bundle ?? "learnings";
    const limit = input.limit ?? LIMIT;

    return forTenant(
      tenantId,
      async (tx) => {
        // As revisões apontam para a ROW do conceito, não para o caminho OKF —
        // o caminho pode ser reusado após um descarte, a row não.
        const [concept] = await tx
          .select({ rowId: concepts.id, type: concepts.type })
          .from(concepts)
          .where(
            and(
              eq(concepts.tenantId, tenantId),
              eq(concepts.bundle, bundle),
              eq(concepts.conceptId, input.conceptId),
            ),
          )
          .limit(1);

        if (concept === undefined) {
          return notFound(
            "conceito_nao_encontrado",
            `Nenhum conceito '${input.conceptId}' no bundle ${bundle}.`,
            { hint: "Liste os conceitos com read_concept (prefix ou type) para achar o caminho certo." },
          );
        }

        const rows = await tx
          .select()
          .from(conceptRevisions)
          .where(
            and(
              eq(conceptRevisions.tenantId, tenantId),
              eq(conceptRevisions.conceptRowId, concept.rowId),
            ),
          )
          .orderBy(desc(conceptRevisions.createdAt))
          .limit(limit + 1);

        const truncated = rows.length > limit;
        const page = truncated ? rows.slice(0, limit) : rows;

        return {
          conceptId: input.conceptId,
          bundle,
          type: concept.type,
          count: page.length,
          ...(truncated
            ? {
                truncated: true as const,
                note: `Mostrando as ${limit} revisões mais recentes; há mais.`,
              }
            : {}),
          revisions: page.map((row, index) => ({
            revisionId: row.id,
            // A mais recente É o estado atual do conceito: toda escrita de
            // save_concept grava a revisão junto.
            ...(index === 0 ? { current: true as const } : {}),
            author: row.author,
            reason: row.reason,
            createdAt: row.createdAt.toISOString(),
            title: row.frontmatter?.title ?? null,
            body: row.body,
          })),
          revert:
            "Para reverter: chame save_concept com o MESMO conceptId e o body/title da revisão escolhida, explicando no reason que é uma reversão. Vira revisão nova; nada é apagado.",
        };
      },
      getDb(),
    );
  },
});

File: services/agent/clara/tools/read_concept_history.py
"""O histórico de um conceito — porta de `agent/tools/read_concept_history.ts`.

`save_concept` sempre prometeu: "reverter é escrever de volta um corpo
antigo". Esta é a metade que faltava: sem um leitor, `concept_revisions`
era append-only sem forma de ver o corpo antigo para escrevê-lo de volta.
"""

from __future__ import annotations

from dataclasses import dataclass, field

from sqlalchemy import select
from sqlalchemy.orm import Session

from clara.db.models import Bundle, Concept, ConceptRevision
from clara.tools.errors import ToolError, not_found

LIMIT = 20

REVERT_HINT = (
    "Para reverter: chame save_concept com o MESMO concept_id e o body/title da revisão "
    "escolhida, explicando no reason que é uma reversão. Vira revisão nova; nada é apagado."
)


@dataclass(frozen=True)
class RevisionSummary:
    revision_id: str
    author: str
    reason: str | None
    created_at: str
    title: str | None
    body: str
    current: bool = False


@dataclass(frozen=True)
class ConceptHistoryResult:
    concept_id: str
    bundle: Bundle
    type: str
    count: int
    revisions: list[RevisionSummary] = field(default_factory=list)
    truncated: bool = False
    note: str | None = None
    revert: str = REVERT_HINT


def read_concept_history(
    session: Session,
    tenant_id: str,
    *,
    concept_id: str,
    bundle: Bundle = "learnings",
    limit: int = LIMIT,
) -> ConceptHistoryResult | ToolError:
    # As revisões apontam para a ROW do conceito, não para o caminho OKF — o
    # caminho pode ser reusado após um descarte, a row não.
    concept = session.execute(
        select(Concept).where(
            Concept.tenant_id == tenant_id,
            Concept.bundle == bundle,
            Concept.concept_id == concept_id,
        )
    ).scalar_one_or_none()

    if concept is None:
        return not_found(
            "conceito_nao_encontrado",
            f"Nenhum conceito '{concept_id}' no bundle {bundle}.",
            hint="Liste os conceitos com read_concept (prefix ou concept_type) para achar o "
            "caminho certo.",
        )

    rows = (
        session.execute(
            select(ConceptRevision)
            .where(
                ConceptRevision.tenant_id == tenant_id,
                ConceptRevision.concept_row_id == concept.id,
            )
            .order_by(ConceptRevision.created_at.desc())
            .limit(limit + 1)
        )
        .scalars()
        .all()
    )

    truncated = len(rows) > limit
    page = rows[:limit] if truncated else rows

    return ConceptHistoryResult(
        concept_id=concept_id,
        bundle=bundle,
        type=concept.type,
        count=len(page),
        truncated=truncated,
        note=f"Mostrando as {limit} revisões mais recentes; há mais." if truncated else None,
        revisions=[
            RevisionSummary(
                revision_id=row.id,
                # A mais recente É o estado atual: toda escrita de
                # save_concept grava a revisão junto.
                current=(index == 0),
                author=row.author,
                reason=row.reason,
                created_at=row.created_at.isoformat(),
                title=row.frontmatter.get("title") if row.frontmatter else None,
                body=row.body,
            )
            for index, row in enumerate(page)
        ],
    )
